File: HW/HW6/0221-HW6.py
from typing import List

def threeSum(nums: List[int]) -> List[List[int]]:
    result = [] #empty list
    array_length = len(nums)    #just for easier referenceing
  
    for i in range(array_length - 2):
        if i > 0 and nums[i] == nums[i - 1]:    #skip duplicate values to avoid duplicate triplates
            continue
        left, right = i + 1, array_length - 1
        while left < right:
            total = nums[i] + nums[left] + nums[right]
            if total == 0:
                result.append([nums[i], nums[left], nums[right]])
                while left < right and nums[left] == nums[left+1]:  #skip duplicates
                    left += 1
                while left < right and nums[right] == nums[right-1]: #skip duplicates
                    right -= 1
                left += 1 #move left pointer right
                right -= 1 #move right pointer left
            elif total < 0: #move left pointer right
                left += 1
            else:   #move right pointer left
                right -= 1
    return result 

testcase1 = [0,1,1]
print(threeSum(testcase1))

testcase2 = [-5, 0, 5, 10, -10, 0]
print(threeSum(testcase2))
#i get duplicate results for this testcase :(

File: HW/HW6/0221-HW6.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ThreeSum {
    public List<List<Integer>> triplet(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();

        for (int i = 0; i < nums.length - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) { continue ;}
            int left = i + 1;
            int right = nums.length - 1;
            while (i + 1 < nums.length - 1) {
                if ((nums[i] + nums[left] + nums[right]) == 0) {
                    result.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    while (left < right && nums[left] == nums[left+1]) left++;
                    while (left < right && nums[right] == nums[right-1]) right++;
                    left++;
                    right--;
                } else if ((nums[i] + nums[left] + nums[right]) < 0) {
                    left++;
                } else right --;

            } //while
        } //for
    } //triplet

    public static void main(String[] args) {
        ThreeSum threesum = new ThreeSum();

        int[] testcase1 = {0, 1, 1};
        System.out.println(threesum.triplet(testcase1));

        int[] testcase2 = {-5, 0, 5, 10, -10, 0};
        System.out.println(threesum.triplet(testcase2));

    } //main
} //threesum